itests/kit/node_unmanaged.go

```diff
@@ -24,3 +60,730 @@ type TestUnmanagedMiner struct {
+			CID:  piece.PieceCID,
+			Size: piece.Size,
+		})
+	}
+
+	r = tm.submitMessage(ctx, &miner14.ProveCommitSectors3Params{
+		SectorActivations:        []miner14.SectorActivationManifest{{SectorNumber: sectorNumber, Pieces: manifest}},
+		SectorProofs:             [][]byte{proveCommit},
+		RequireActivationSuccess: true,
+	}, 1, builtin.MethodsMiner.ProveCommitSectors3)
+	req.True(r.Receipt.ExitCode.IsSuccess())
+
+	tm.proofType[sectorNumber] = proofType
+
+	respCh := make(chan WindowPostResp, 1)
+
+	go tm.wdPostLoop(ctx, sectorNumber, respCh)
+
+	return sectorNumber, respCh
+}
+
+func (tm *TestUnmanagedMiner) OnboardCCSectorWithRealProofs(ctx context.Context, proofType abi.RegisteredSealProof) (abi.SectorNumber, chan WindowPostResp) {
+	req := require.New(tm.t)
+	sectorNumber := tm.currentSectorNum
+	tm.currentSectorNum++
+
+	// --------------------Create pre-commit for the CC sector -> we'll just pre-commit `sector size` worth of 0s for this CC sector
+
+	// Step 1: Wait for the pre-commitseal randomness to be available (we can only draw seal randomness from tipsets that have already achieved finality)
+	preCommitSealRand := tm.waitPreCommitSealRandomness(ctx, sectorNumber)
+
+	// Step 2: Write empty bytes that we want to seal i.e. create our CC sector
+	tm.makeAndSaveCCSector(ctx, sectorNumber)
+
+	// Step 3: Generate a Pre-Commit for the CC sector -> this persists the proof on the `TestUnmanagedMiner` Miner State
+	tm.generatePreCommit(ctx, sectorNumber, preCommitSealRand, proofType, []abi.PieceInfo{})
+
+	// Step 4 : Submit the Pre-Commit to the network
+	r := tm.submitMessage(ctx, &miner14.PreCommitSectorBatchParams2{
+		Sectors: []miner14.SectorPreCommitInfo{{
+			Expiration:    2880 * 300,
+			SectorNumber:  sectorNumber,
+			SealProof:     TestSpt,
+			SealedCID:     tm.sealedCids[sectorNumber],
+			SealRandEpoch: preCommitSealRand,
+		}},
+	}, 1, builtin.MethodsMiner.PreCommitSectorBatch2)
+	req.True(r.Receipt.ExitCode.IsSuccess())
+
+	// Step 5: Generate a ProveCommit for the CC sector
+	waitSeedRandomness := tm.proveCommitWaitSeed(ctx, sectorNumber)
+
+	proveCommit := tm.generateProveCommit(ctx, sectorNumber, proofType, waitSeedRandomness, []abi.PieceInfo{})
+
+	// Step 6: Submit the ProveCommit to the network
+	tm.t.Log("Submitting ProveCommitSector ...")
+
+	r = tm.submitMessage(ctx, &miner14.ProveCommitSectors3Params{
+		SectorActivations:        []miner14.SectorActivationManifest{{SectorNumber: sectorNumber}},
+		SectorProofs:             [][]byte{proveCommit},
+		RequireActivationSuccess: true,
+	}, 0, builtin.MethodsMiner.ProveCommitSectors3)
+	req.True(r.Receipt.ExitCode.IsSuccess())
+
+	tm.proofType[sectorNumber] = proofType
+
+	respCh := make(chan WindowPostResp, 1)
+
+	go tm.wdPostLoop(ctx, sectorNumber, respCh)
+
+	return sectorNumber, respCh
+}
+
+func (tm *TestUnmanagedMiner) wdPostLoop(ctx context.Context, sectorNumber abi.SectorNumber, respCh chan WindowPostResp) {
+	go func() {
```

Comment:
This goroutine is using `*testing.T`, which isn't safe unless you're only using `t.Error()`, which it's not (I think `t.Log*` is OK though). For a start, `tm.FullNode.WaitTillChain` uses `require.NoError` and all the `require.*` methods end in `t.FailNow` which isn't OK. The `assert.*` variants don't `t.FailNow` afaik though.

See https://pkg.go.dev/testing#T.FailNow

That's why I had all that awkward infrastructure to (a) send errors back to the main goroutine (although that may not have been necessary if it just used `assert.NoError` and friends, or just `t.Error` and then get the main thread to fail the test with channel-closed signal or something like that) and (b) increment the chain manually without `WaitTillChain`.

Reply:
Fixed this by creating a `WaitTillChainOrError` function that propagates the error back to the caller